### ISSUE TITLE
phpcs.xml properties 'file' and 'exclude-pattern' are not handled right

### ISSUE THREAD

**Issue by acalvino4** [squizlabs/PHP_CodeSniffer#3438]
**Describe the bug**
phpcs does not exclude files for sniffing as documented in the [annotated rule set](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-Ruleset), specifically with the `<exclude-pattern>` property.

I created a `phpcs.xml` in my project root directory with the contents below, and then ran `phpcs` from the project root. Errors were reported both for files outside the `<file>` directory, and from within the `<exclude-pattern>` directories.

**Custom ruleset**
```xml
<?xml version="1.0"?>
<ruleset name="A Custom Standard">

  <description>A Template for extending from PSR12</description>

  <file>./system/user</file>
  <exclude-pattern type="relative">./system/user/addons/bloqs/*</exclude-pattern>
  <exclude-pattern type="relative">./system/user/addons/wygwam/*</exclude-pattern>

  <rule ref="PSR12">
    <!-- Exclude PSR12 Rules Here -->
  </rule>

  <!-- Add Additional Rules Here -->

</ruleset>
```

**To reproduce**
1. Create a file called `phpcs.xml` with the code sample above in your project root
1. Run `phpcs` from your project root
1. See error output from files that shouldn't be included
```
...

FILE: /var/www/site/system/user/addons/wygwam/language/english/wygwam_lang.php
----------------------------------------------------------------------------------------------------------------------
FOUND 3 ERRORS AND 4 WARNINGS AFFECTING 6 LINES
----------------------------------------------------------------------------------------------------------------------
  1 | ERROR   | [x] Header blocks must be separated by a single blank line
    |         |     (PSR12.Files.FileHeader.SpacingAfterBlock)
 23 | WARNING | [ ] Line exceeds 120 characters; contains 126 characters (Generic.Files.LineLength.TooLong)

<errors for other wygwam and bloq files>
...
```

**Expected behavior**
The files above should have no output because they are explicitly excluded. 

**Versions (please complete the following information):**
 - OS: Ubuntu 20.04.3 LTS
 - PHP: 8.0.10
 - PHPCS: 3.6.0
 - Standard: Extension of PSR12

**Additional Context**
I attempted getting this to work with a variety of glob patterns, to no avail.

**Comment by acalvino4** [comment]
After further testing, seems like the file directive is indeed working correctly. The exclude-pattern portion I have still not gotten to work though. I will edit the original.

**Comment by jrfnl** [comment]
@acalvino4 Could you please have a try with the below ? and let me know whether that works for you ?

```xml
  <file>./system/user</file>
  <exclude-pattern type="relative">^/addons/bloqs/*</exclude-pattern>
  <exclude-pattern type="relative">^/addons/wygwam/*</exclude-pattern>
```

As per the example in the wiki:
> The relative path is determined based on the paths you pass into PHP_CodeSniffer on the command line.

In this case, that means the path you pass to `<file>`.

**Comment by acalvino4** [comment]
@jrfnl thanks for the response. I did notice that on the wiki page, and tried that as well, but it doesn't work either :/

If this just isn't supported, I think this could be considered an enhancement request; specifically to allow excluding patterns relative to project root / phpcs.xml file location, as that would be the most straightforward.

I was able to get it working via the non-relative syntax, but the issue with that is if I have multiple folders of the name I am trying to exclude, and only want some of them excluded, it will ignore all.

**Comment by jrfnl** [comment]
@acalvino4 I'm sorry that didn't work. I know from my own experience that the relative path bit can be a bit fiddly to get right/working. 

> I was able to get it working via the non-relative syntax, but the issue with that is if I have multiple folders of the name I am trying to exclude, and only want some of them excluded, it will ignore all.

Ah, but that might be solvable. It is a little known fact that `<exclude-pattern>`s are basically [PCRE regexes](https://www.php.net/manual/en/reference.pcre.pattern.syntax.php) with automatic replacement of `*` with `.*`.

It is [mentioned in the wiki](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#ignoring-files-and-folders), but easy to miss:

> The ignore patterns are treated as regular expressions. If you do specify a regular expression, be aware that * is converted to .* for the convenience in simple patterns, like those used in the example above. So use * anywhere you would normally use .*. Also ensure you escape any . characters that you want treated as a literal dot, such as when checking file extensions. So if you are checking for .inc in your ignore pattern, use \.inc instead.

So, you could use a pattern like this:
```xml
  <exclude-pattern>/system/user/addons/(?:bloqs|wygwam)/*\.php$</exclude-pattern>
```

Not sure if that is the right pattern for your repo, but all PCRE features are available, including look aheads/behinds, so I'm sure you can probably come up with something which would work for the repo.

Hope that helps.

**Comment by acalvino4** [comment]
Thanks a lot for the info - I'll give that a shot!

**Comment by jrfnl** [comment]
@acalvino4 Just checking in- did you get it working ? If so, can this issue be closed ?

**Comment by acalvino4** [comment]
It did work! thanks for the tip.

**Comment by mreduar** [comment]
This is a problem when you have directories with the same name in your project.
